arangod/Pregel/Actor/MPSCQueue.h

```diff
@@ -0,0 +1,96 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Markus Pfeiffer
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <memory>
+#include <atomic>
+
+// This lock-free multi-producer-single-consumer-queue is inspired by the
+// code published at
+//
+// https://www.1024cores.net/home/lock-free-algorithms/queues/intrusive-mpsc-node-based-queue
+//
+// under a Apache-2.0 license.
+namespace arangodb::pregel::mpscqueue {
+
+struct MPSCQueue {
+  struct Node {
+    virtual ~Node() = default;
+    std::atomic<Node*> next{nullptr};
+  };
+
+  MPSCQueue() : stub{}, head(&stub), tail(&stub) {}
+
+  auto push_internal(Node *value) -> void {
+    auto prev = head.exchange(value);
+    prev->next.store(value);
```

Comment:
What if between line 46 and 47, `prev` was popped and released? Or even worse, it was pop, released, reallocated and pushed somewhere else. Now you write to some random memory address. I think this is the classical ABA problem and memory reclamation problem. I have no idea, so maybe ask @mpoeter .

Reply:
This should not happen (according to my current understanding)

Since you're inserting at the head, the prev's old next pointer would be a nullptr (before you store something in it);

The pop function then would not pop directly, but add a pointer to stub exactly to avoid this situation.

Reply:
Correct - the node referenced by head can never be released/reclaimed as long as the next pointer is null. That is the beauty of this queue, because it does not need a memory reclamation scheme.